Marlin/src/lcd/dogm/u8g_dev_st7920_128_64_sw_spi.cpp

```diff
@@ -0,0 +1,277 @@
+ * Copyright (C) 2016 MarlinFirmware [https://github.com/MarlinFirmware/Marlin]
+ *
+ * Based on Sprinter and grbl.
+ * Copyright (C) 2011 Camiel Gubbels / Erik van der Zalm
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+/*
+ * Please, note: 
+ *  We need our custom implementation for Software SPI, as the default implementation 
+ * of U8GLIB, when running in an ARM based board, is too fast and the display will not
+ * recognize commands and/or data at such speeds. This implementation autoderives the
+ * required delays to get the maximum possible performance by using the F_CPU macro that
+ * specifies the CPU speed. According to the ST7920 datasheet, the maximum SCLK is 1MHz.
+ */
+ 
+#ifndef ULCDST7920_SWSPI_H
+#define ULCDST7920_SWSPI_H
+
+#include "../../Marlin.h"
+
+#if ENABLED(U8GLIB_ST7920)
+
+#include <U8glib.h>
+
+#define ST7920_CLK_PIN  LCD_PINS_D4
+#define ST7920_DAT_PIN  LCD_PINS_ENABLE
+#define ST7920_CS_PIN   LCD_PINS_RS
+
+//#define PAGE_HEIGHT 8   //128 byte framebuffer
+#define PAGE_HEIGHT 16  	//256 byte framebuffer
+//#define PAGE_HEIGHT 32  //512 byte framebuffer
+
+#define LCD_PIXEL_WIDTH 128
+#define LCD_PIXEL_HEIGHT 64
+
+//set optimization so ARDUINO optimizes this file
+#pragma GCC optimize (3)
+
+/* ---------------- Delay Cycles routine -------------- */
+
+#if defined(__arm__)
+/* https://blueprints.launchpad.net/gcc-arm-embedded/+spec/delay-cycles */
+
+#define nop() __asm__ __volatile__("nop;\n\t":::)
+
+static __inline__ __attribute__((__always_inline__)) void __delay_4cycles(uint32_t cy) // +1 cycle
+{
+ #if ARCH_PIPELINE_RELOAD_CYCLES<2
+ #	define EXTRA_NOP_CYCLES "nop"
+ #else
+ #	define EXTRA_NOP_CYCLES ""
+ #endif
+
+ __asm__ __volatile__
+ (
+  ".syntax unified" "\n\t" // is to prevent CM0,CM1 non-unified syntax
+	
+  "loop%=:" "\n\t"
+  "	subs %[cnt],#1" "\n\t"
+  EXTRA_NOP_CYCLES "\n\t"
+  " bne loop%=" "\n\t"
+  : [cnt]"+r"(cy) // output: +r means input+output
+  : // input:
+  : "cc" // clobbers:
+ );
+}
+
+static __inline__ __attribute__((__always_inline__)) void DELAY_CYCLES(uint32_t x)
+{
+  if (__builtin_constant_p(x)) {
+#define MAXNOPS 4
+
+		if (x<=MAXNOPS) {
+			if (x==1) {nop();}
+			else if (x==2) {nop(); nop();}
+			else if (x==3) {nop(); nop(); nop();}
+			else if (x==4) {nop(); nop(); nop(); nop();}
+		} else { // because of +1 cycle inside delay_4cycles
```

Comment:
If code from 3rd-party libraries has terrible formatting and tabs, we can leave it. But new Marlin-specific code should follow Marlin formatting rules.

Reply:
I'll reformat it as needed. What i'd like to know if there are any suggestions on rearranging things. And one of those things is exactly that function you have pointed out ;)... The DELAY_CYCLES is a very generic one, usable for all ARM variants, and also for AVR ... Maybe muve it to a better place ?

Reply:
And regarding to the serial port numbers, Arduino DUE has 2 serial ports: One of them is implemented by using a USB to Serial converter that interfaces to the RX/TX pins on them SAM3X - That one is the native port that the SAM-BA bootloader uses.,  The other one is implemented by using the SAM3X native USB port and implementing a CDC class device.
SERIAL_PORT=-1 would mean to use that last one, SERIAL_PORT=0 means to use the USB to serial converter chip... At least, that is exactlyhow this patch leaves things

Reply:
For multi-host we were trying to stay away from any system that would add more overhead to AVR. If there are going to be multiple instances of any class —requiring instance-based calling conventions— then it should only apply to the 32-bit platforms.

Reply:
But, let's keep those mods to the serial port class in a separate pull request right now (Y)